crates/ty_python_semantic/src/types.rs

```diff
@@ -1030,7 +1036,64 @@ impl<'db> Type<'db> {
             return None;
         }
 
-        specialization
+        Some((class_type, specialization?))
+    }
+
+    /// Given a type variable `T` from the generic context of a class `C`:
+    /// - If `self` is a specialized instance of `C`, returns the type assigned to `T` on `self`.
+    /// - If `self` is a specialized instance of some class `A`, and `C` is a subclass of `A`
+    ///   such that the type variable `U` on `A` is specialized to `T`, returns the type
+    ///   assigned to `U` on `self`.
+    pub(crate) fn find_type_var_from(
+        self,
+        db: &'db dyn Db,
+        bound_typevar: BoundTypeVarInstance<'db>,
+        class: ClassLiteral<'db>,
+    ) -> Option<Type<'db>> {
+        self.find_type_var_from_impl(db, bound_typevar, class, &mut FxHashSet::default())
+    }
+
+    pub(crate) fn find_type_var_from_impl(
+        self,
+        db: &'db dyn Db,
+        bound_typevar: BoundTypeVarInstance<'db>,
+        class: ClassLiteral<'db>,
+        visited: &mut FxHashSet<(ClassLiteral<'db>, BoundTypeVarIdentity<'db>)>,
+    ) -> Option<Type<'db>> {
+        if let Some(specialization) = self.specialization_of(db, class) {
+            return specialization.get(db, bound_typevar);
+        }
+
+        // TODO: We should use the constraint solver here to determine the type mappings for more
+        // complex subtyping relationships, e.g., `type[C[T]]` to `Callable[..., T]`, or unions
+        // containing multiple generic elements.
+        for base in class.iter_mro(db, None) {
+            let Some((origin, Some(specialization))) =
+                base.into_class().map(|class| class.class_literal(db))
+            else {
+                continue;
+            };
+
+            for (base_typevar, base_ty) in specialization
+                .generic_context(db)
+                .variables(db)
+                .zip(specialization.types(db))
+            {
+                if *base_ty == Type::TypeVar(bound_typevar) {
+                    if !visited.insert((origin, base_typevar.identity(db))) {
+                        return None;
+                    }
+
+                    if let Some(ty) =
+                        self.find_type_var_from_impl(db, base_typevar, origin, visited)
+                    {
+                        return Some(ty);
+                    }
+                }
+            }
```

Comment:
we do a similar loop through the MRO in the (old) constraint solver which doesn't seem to need to do the `FxHashSet` memoization thing -- could you also take a similar approach here? Might that be more performant?

https://github.com/astral-sh/ruff/blob/2250fa6f98f4fba4aaf17ff0a5e4f0be24419986/crates/ty_python_semantic/src/types/generics.rs#L1598-L1641